src/server/scripts/Northrend/CrusadersColiseum/TrialOfTheChampion/boss_argent_challenge.cpp

```diff
@@ -104,321 +150,710 @@ enum Spells
-                DoCastAOE(SPELL_HOLY_NOVA, false);
-                DoCast(me, SPELL_SHIELD);
-                DoCastAOE(SPELL_SUMMON_MEMORY, false);
-                DoCastAOE(SPELL_CONFESS, false);
+                _shielded = true;
+                damage = me->GetHealth() - 1;
+                me->AttackStop();
+
+                DoCast(SPELL_DIVINE_SHIELD);
+                DoCast(SPELL_FINAL_MEDITATION);
+            }
+        }
+
+        void JustDied(Unit* /*killer*/) override
+        {
+            _instance->SetData(DATA_ARGENT_SOLDIER_DEFEATED, _instance->GetData(DATA_ARGENT_SOLDIER_DEFEATED) + 1);
+            if (me->GetEntry() == NPC_PRIESTESS)
+            {
+                if (Creature* fountain = ObjectAccessor::GetCreature(*me, _fountainGuid))
+                    fountain->DisappearAndDie();
+            }
+        }
 
-                bHealth = true;
+        void SetData(uint32 type, uint32 /*data*/) override
+        {
+            _waypointId = type;
+            Position homePos;
+            GenerateFinalPosition(homePos);
+
+            if (homePos.IsPositionValid())
+            {
+                me->SetHomePosition(homePos);
+                me->SetWalk(true);
+                if (me->GetEntry() == NPC_ARGENT_MONK)
+                    me->GetMotionMaster()->MovePoint(POINT_PREFIGHT, homePos);
+            }
+        }
+
+        void JustEngagedWith(Unit* /*who*/) override
+        {
+            switch (me->GetEntry())
+            {
+            case NPC_ARGENT_LIGHWIELDER:
+                _events.ScheduleEvent(EVENT_BLAZING_LIGHT, 10 * IN_MILLISECONDS);
+                _events.ScheduleEvent(EVENT_CLEAVE, urand(4 * IN_MILLISECONDS, 6 * IN_MILLISECONDS));
+                if (IsHeroic())
+                    _events.ScheduleEvent(EVENT_UNBALANCING_STRIKE, 8 * IN_MILLISECONDS);
+                break;
+            case NPC_ARGENT_MONK:
+                _events.ScheduleEvent(EVENT_FLURRY_OF_BLOWS, 2 * IN_MILLISECONDS);
+                _events.ScheduleEvent(EVENT_PUMMEL, 12 * IN_MILLISECONDS);
+                break;
+            case NPC_PRIESTESS:
+                _events.ScheduleEvent(EVENT_SHADOW_WORD_PAIN, 500);
+                _events.ScheduleEvent(EVENT_HOLY_SMITE, 2500);
+                _events.ScheduleEvent(EVENT_FOUNTAIN, 10 * IN_MILLISECONDS);
+                if (IsHeroic())
+                    _events.ScheduleEvent(EVENT_MIND_CONTROL, 15 * IN_MILLISECONDS);
+                break;
+            default:
+                break;
+            }
+        }
+
+        void AttackStart(Unit* who) override
+        {
+            if (!(me->GetEntry() == NPC_ARGENT_MONK && me->HasAura(SPELL_DIVINE_SHIELD)))
+                ScriptedAI::AttackStart(who);
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            _events.Update(diff);
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (!UpdateVictim())
```

Comment:
why is this after the eventmap invocation and casting state check? should npcs not evade while casting?